File: dataforge-control/src/main/java/hep/dataforge/control/ports/VirtualPort.java
/* 
 * Copyright 2015 Alexander Nozik.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package hep.dataforge.control.ports;

import hep.dataforge.exceptions.PortException;
import hep.dataforge.meta.Configurable;
import hep.dataforge.meta.Configuration;
import hep.dataforge.meta.Meta;

import java.time.Duration;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.*;
import java.util.function.Supplier;

/**
 * @author Alexander Nozik
 */
public abstract class VirtualPort extends PortHandler implements Configurable {

    private final Set<TaggedFuture> futures = new CopyOnWriteArraySet<>();
    private ScheduledExecutorService scheduler;
    private boolean isOpen = false;
    private Configuration configuration = new Configuration("virtualPort");

    @Override
    public void open() throws PortException {
        scheduler = Executors.newScheduledThreadPool(meta().getInt("numThreads", 4));
        isOpen = true;
    }

    @Override
    public boolean isOpen() {
        return isOpen;
    }

    @Override
    public Configuration getConfig() {
        return configuration;
    }

    @Override
    public Configurable configure(Meta config) {
        configuration.update(config);
        return this;
    }

    @Override
    public Meta meta() {
        return configuration;
    }

    @Override
    public String getPortId() {
        return meta().getString("id", getClass().getSimpleName());
    }

    @Override
    public void send(String message) throws PortException {
        evaluateRequest(message);
    }

    /**
     * The device logic here
     *
     * @param request
     */
    protected abstract void evaluateRequest(String request);

    protected synchronized void clearCompleted() {
        futures.stream().filter((future) -> (future.future.isDone() || future.future.isCancelled())).forEach(futures::remove);
    }

    protected synchronized void cancelByTag(String tag) {
        futures.stream().filter((future) -> future.hasTag(tag)).forEach(TaggedFuture::cancel);
    }

    /**
     * Plan the response with given delay
     *
     * @param response
     * @param delay
     * @param tags
     */
    protected synchronized void planResponse(String response, Duration delay, String... tags) {
        clearCompleted();
        Runnable task = () -> receivePhrase(response);
        ScheduledFuture future = scheduler.schedule(task, delay.toNanos(), TimeUnit.NANOSECONDS);
        this.futures.add(new TaggedFuture(future, tags));
    }

    protected synchronized void planRegularResponse(Supplier<String> responseBuilder, Duration delay, Duration period, String... tags) {
        clearCompleted();
        Runnable task = () -> receivePhrase(responseBuilder.get());
        ScheduledFuture future = scheduler.scheduleAtFixedRate(task, delay.toNanos(), period.toNanos(), TimeUnit.NANOSECONDS);
        this.futures.add(new TaggedFuture(future, tags));
    }

    @Override
    public void close() throws Exception {
        futures.clear();
        this.scheduler.shutdownNow();
        isOpen = false;
    }

    private class TaggedFuture {

        public TaggedFuture(ScheduledFuture future, String... tags) {
            this.future = future;
            this.tags = new HashSet<>();
            this.tags.addAll(Arrays.asList(tags));
        }

        ScheduledFuture future;
        Set<String> tags;

        public boolean hasTag(String tag) {
            return tags.contains(tag);
        }

        public boolean cancel() {
            return future.cancel(true);
        }
    }
}

File: dataforge-control/src/main/kotlin/hep/dataforge/control/ports/VirtualPort.kt
/*
 * Copyright  2017 Alexander Nozik.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package hep.dataforge.control.ports

import hep.dataforge.exceptions.PortException
import hep.dataforge.meta.Configurable
import hep.dataforge.meta.Configuration
import hep.dataforge.meta.Meta
import java.time.Duration
import java.util.*
import java.util.concurrent.*
import java.util.function.Supplier

/**
 * @author Alexander Nozik
 */
abstract class VirtualPort protected constructor(meta: Meta) : Port(meta), Configurable {

    private val futures = CopyOnWriteArraySet<TaggedFuture>()
    private var scheduler: ScheduledExecutorService? = null
    override var isOpen = false
    private var configuration = Configuration("virtualPort")

    init {
        configuration = Configuration(meta)
    }

    @Throws(PortException::class)
    override fun open() {
        scheduler = Executors.newScheduledThreadPool(meta.getInt("numThreads", 4)!!)
        isOpen = true
    }

    override fun getConfig(): Configuration {
        return configuration
    }

    override fun configure(config: Meta): Configurable {
        configuration.update(config)
        return this
    }

    override fun getMeta(): Meta {
        return configuration
    }

    override fun toString(): String {
        return meta.getString("id", javaClass.simpleName)
    }

    @Throws(PortException::class)
    public override fun send(message: String) {
        evaluateRequest(message)
    }

    /**
     * The device logic here
     *
     * @param request
     */
    protected abstract fun evaluateRequest(request: String)

    @Synchronized protected fun clearCompleted() {
        futures.stream().filter { future -> future.future.isDone || future.future.isCancelled }.forEach{ futures.remove(it) }
    }

    @Synchronized protected fun cancelByTag(tag: String) {
        futures.stream().filter { future -> future.hasTag(tag) }.forEach{ it.cancel() }
    }

    /**
     * Plan the response with given delay
     *
     * @param response
     * @param delay
     * @param tags
     */
    @Synchronized protected fun planResponse(response: String, delay: Duration, vararg tags: String) {
        clearCompleted()
        val task = { receivePhrase(response) }
        val future = scheduler!!.schedule(task, delay.toNanos(), TimeUnit.NANOSECONDS)
        this.futures.add(TaggedFuture(future, *tags))
    }

    @Synchronized protected fun planRegularResponse(responseBuilder: Supplier<String>, delay: Duration, period: Duration, vararg tags: String) {
        clearCompleted()
        val task = { receivePhrase(responseBuilder.get()) }
        val future = scheduler!!.scheduleAtFixedRate(task, delay.toNanos(), period.toNanos(), TimeUnit.NANOSECONDS)
        this.futures.add(TaggedFuture(future, *tags))
    }

    @Throws(Exception::class)
    override fun close() {
        futures.clear()
        this.scheduler!!.shutdownNow()
        isOpen = false
        super.close()
    }

    private inner class TaggedFuture(internal var future: ScheduledFuture<*>, vararg tags: String) {
        internal var tags: MutableSet<String>

        init {
            this.tags = HashSet()
            this.tags.addAll(Arrays.asList(*tags))
        }

        fun hasTag(tag: String): Boolean {
            return tags.contains(tag)
        }

        fun cancel(): Boolean {
            return future.cancel(true)
        }
    }
}
